### ISSUE TITLE
Ternary equality / between comparator

### ISSUE THREAD

**Issue by EBrown8534** [fsharp/fslang-suggestions#676]
# Ternary equality / between comparator

I propose we add support for the "between" ternary syntax: `a < x < b`. This should support `<` and `<=`.

The existing way of approaching this problem in F# is to write a multi-expression comparison: `a < x && x < b`.

## Pros and Cons 

The advantages of making this adjustment to F# are a more succinct and legible syntax for testing a value between two other values.

The disadvantages of making this adjustment to F# are:

* I'm pretty sure it is a lot of work, F# currently has no ternary expressions that I'm aware of, so this is a brand new concept to the specification;
* It's specific to `<` or `<=`, so it's only beneficial in highly limited cases;
* The benefits are not substantial, and overloading the operator would become highly complicated;

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): L/XL

Related suggestions: https://github.com/fsharp/fslang-suggestions/issues/675

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

**Comment by abelbraaksma** [comment]
> This is not a breaking change to the F# language design

I'm not sure this applies. Taking your suggestion literally (that is, without suggesting updated syntax), the following will lead to a different result:

```f#
true > false && false > false  // original syntax for "between", yields false
true > false > false           // new syntax for "between" currently yields true, but should then become false
```

The other thing about "breaking change" being the one you already mentioned: ternary operators are now not allowed, so it will require careful thinking finding out whether chained operators could now suddenly be redefined as ternary (think `x |> y |> z`). One way forward might be to require a special syntax or designator for ternary operators.

Note also that ternary operators are rare, most languages (though notably not F#) support only the `.. ? .. : ..` ternary conditional operator.

**Comment by EBrown8534** [comment]
I don't disagree at all.

As a regular, non-expert, user of F#, I think this whole idea is "meh", I don't think it's something that needs to make it in the spec. (I mostly proposed it so that we can have the discussion on it now, and get it out of the way.) Being on the low-end of the expertise spectrum, it really wouldn't convince me all that much to use it.

While it adds a little value (arguable), some of the situations you mentioned to create problems. In Slack, Diego Esmerio (@Nhowka) mentioned the following active pattern that does the same thing:

>     let (|Between|_|) a b x =
>       let m = min a b
>       let n = max a b
>       if  x > m && x < n then
>         Some Between
>       else
>         None
> 
>     let test = function
>       | Between 0 5 -> printfn "Passed"
>       | _ -> printfn "Outside range"
> 
>     test 4
>     test 8

Personally, I think this would be a far better feature to include in the core F# spec, and would alleviate both issues much more cleanly and idiomatically. With this pattern, I wouldn't need to have `| 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10`, I can use `Between 0 11`. (Maybe we make it inclusive, instead of exclusive, for better support of ranges.)

**Comment by abelbraaksma** [comment]
Yes, it is trivial to add the `Between` active pattern. Though when adding it to FSharp.Core we'll need to answer things like "is it inclusive, exclusive, left-inclusive?" and "How do we treat NaN?"

As an alternative, have a look here, it has two operators defined that act as 'between": http://www.fssnip.net/9y/title/Between-operators

**Comment by EBrown8534** [comment]
@abelbraaksma I was actually thinking about that this morning, because I added it to my own extensions library. I went with two items:

* [`Between a b`](https://github.com/EBrown8534/FSharpExtensions/blob/master/FSharpExtensions/Patterns/Between.fs): matches if the value is _inclusively_ in the range (`x >= a && x <= b`);
* [`BetweenExc a b`](https://github.com/EBrown8534/FSharpExtensions/blob/master/FSharpExtensions/Patterns/BetweenExc.fs): matches if the value is _exclusively_ in the range (`x > a && x < b`);

But, alas, it still doesn't answer the `+/- INF` or `NaN` questions, really. It accepts however `>=` / `<=` will treat them. It also complicates usage a bit, but _at least_ it helps with my use-case, and allows me to do what I wanted without complicating things any further than necessary.

I'd prefer to see this type of a pattern-match capability built-in once we nail those questions down, because I think it's immensely more useful in this scenario.

**Comment by cartermp** [comment]
Since this is a breaking change and doesn't seem to dramatically improve the lives of most F# programmers, I doubt this would be taken.

**Comment by EBrown8534** [comment]
@cartermp To be fair, the only argument I would make the opposite is that F# is a language of conveniences, no matter how small. Something like this adds _another_ convenience to the language. (Not that such a reason justifies it.)

I'm on the fence about this anymore, I think it would make the syntax for such things clearer, but it's not something I use (or would) a lot anyway.
